### PR TITLE
[FLINK-30210] Refactor StoreCompactOperator to accept records containing partitions and buckets in Table Store

### PR DESCRIPTION
**Pull request by tsreaper**
In this ticket we refactor StoreCompactOperator to accept records containing partitions and buckets in Table Store. The old ALTER TABLE COMPACT will also be disabled in this ticket due to various restrictions.

### CODE REVIEW

File: flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/StoreSinkWrite.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import java.io.Serializable;
+import java.util.List;
+
+/** Helper class of {@link StoreWriteOperator} for different types of table store sinks. */
+interface StoreSinkWrite {
+
+    @Nullable
+    TableWrite getTableWrite();
+
+    void touchBucket(BinaryRowData partition, int bucket);
```
Comment (JingsongLi) [inline flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/StoreSinkWrite.java:39]: Can we just provide:
```
SinkRecord write(RowData rowData) throws Exception;

void compact(BinaryRowData partition, int bucket, boolean fullCompaction) throws Exception;
```

We don't need this `touchBucket`. `IOManager` and `Overwrite` can be prepared in `Provider.provide`.

File: flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/StoreCompactOperator.java [hunk 2]
```diff
@@ -18,99 +18,70 @@
...
-import javax.annotation.Nullable;
+import org.apache.flink.util.Preconditions;
 
 import java.io.IOException;
 import java.util.List;
-import java.util.Map;
-import java.util.Objects;
-import java.util.stream.Collectors;
 
 /** A dedicated operator for manual triggered compaction. */
```
Comment (JingsongLi) [inline flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/StoreCompactOperator.java:34]: Add comments: Records from where.

File: flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/StoreCompactOperator.java [hunk 3]
```diff
@@ -18,99 +18,70 @@
...
+    private transient RowDataSerializer partitionSerializer;
+    private transient OffsetRowData reusedPartition;
 
     public StoreCompactOperator(
             FileStoreTable table,
-            String commitUser,
-            @Nullable Map<String, String> compactPartitionSpec) {
+            StoreSinkWrite.Provider storeSinkWriteProvider,
+            boolean isStreaming) {
+        Preconditions.checkArgument(
```
Comment (JingsongLi) [inline flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/StoreCompactOperator.java:49]: Does this expected? The user manually starts a compaction job. Does user need to set the option?

Comment (tsreaper) [reply flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/StoreCompactOperator.java:49]: We'll introduce a `TableStoreCompactJob` so user can submit this job class to Flink. This job class will disable "write.compaction-skip" so user does not need to care about it.

This `checkArgument` is just for sanity check.

File: flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/StoreSinkWriteImpl.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        // each job can only have one user name and this name must be consistent across restarts
+        // we cannot use job id as commit user name here because user may change job id by creating
+        // a savepoint, stop the job and then resume from savepoint
+        commitUser =
+                StateUtils.getSingleValueFromState(
+                        context, "commit_user_state", String.class, initialCommitUser);
+
+        // see comments of StateUtils.getSingleValueFromState for why commitUser may be null
+        if (commitUser == null) {
+            write = null;
```
[the remark this answers is not in the document]
Comment (tsreaper) [reply flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/StoreSinkWriteImpl.java:55]: More comments added.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**